README.md
# react-progress-button

> Simple [React](http://facebook.github.io/react/index.html) component for a circular Progress Button.

### [Demo](https://mathieudutour.github.io/react-progress-button)

[![Demo](https://cdn.rawgit.com/mathieudutour/react-progress-button/master/example/demo.gif "Demo")](https://github.com/mathieudutour/react-progress-button/blob/master/example/index.html)

## Install

```bash
npm install react-progress-button --save
```

or

```bash
bower install react-progress-button --save
```

## Example

Controlled usage:

```javascript
var ProgressButton = require('react-progress-button');

var App = React.createClass({
  render() {
    return (
      <div>
        <ProgressButton ref='button' onClick={this.handleClick}>
          Go!
        </ProgressButton>
      </div>
    );
  },
  
  handleClick() {
    this.refs.button.loading();
    //make asynchronious call
    setTimeout(function() {
      this.refs.button.success();
    }.bind(this), 3000);
  }
});
```

## API

### Props

All props are optional.

##### classNamespace

Namespace for CSS classes, default is `pb-` i.e CSS classes are `pb-button`.

##### durationError

Duration (ms) before going back to normal state when an error occurs,
default is 1200

##### durationSuccess

Duration (ms) before going back to normal state when an success occurs,
default is 500

##### onClick

Function to call onClick on the button

##### onError

Function to call when going back to the normal state after an error

##### onSuccess

Function to call when going back to the normal state after a success

##### state

State of the button if you do not want to use the functions. Can be '', 'loading', 'success' or 'error'.


### Methods

##### loading()

Put the button in the loading state.

##### notLoading()

Put the button in the normal state.

##### success([callback])

Put the button in the success state. Call the callback or the onSuccess prop when going back to the normal state.

##### error([callback])

Put the button in the error state. Call the callback or the onSuccess prop when going back to the normal state.

## Styles

Look at [react-progress-button.css](https://github.com/mathieudutour/react-progress-button/blob/master/react-progress-button.css) for an idea on how to style this component.

---

MIT Licensed





CHANGELOG.md
### 0.1.0 (2015-27-04)

* first release


react-progress-button.js
;(function (root, factory) {
  if (typeof module !== "undefined" && module.exports) {
    module.exports = factory(require("react"));
  } else if (typeof define === "function" && define.amd) {
    define(["react"], factory);
  } else {
    root.ProgressButton = factory(root.React);
  }
})(this, function (React) {
  "use strict";

  var ProgressButton = React.createClass({displayName: "ProgressButton",
    propTypes: {
      classNamespace: React.PropTypes.string,
      durationError: React.PropTypes.number,
      durationSuccess: React.PropTypes.number,
      form: React.PropTypes.string,
      onClick: React.PropTypes.func,
      onError: React.PropTypes.func,
      onSuccess: React.PropTypes.func,
      state: React.PropTypes.string,
      type: React.PropTypes.string
    },

    getDefaultProps: function() {
      return {
        classNamespace: 'pb-',
        durationError: 1200,
        durationSuccess: 500,
        state: '',
        onClick: function() {}
      };
    },

    getInitialState: function() {
      return {
        currentState: this.props.state
      };
    },

    render: function() {
      return (
        React.createElement("div", {className: this.props.classNamespace + "container " + this.state.currentState,
          onClick: this.props.onClick},
          React.createElement("button", {type: this.props.type, form: this.props.form,
            className: this.props.classNamespace + "button"},
            React.createElement("span", null, this.props.children),
            React.createElement("svg", {className: this.props.classNamespace + "progress-circle",
              viewBox: "0 0 41 41"},
              React.createElement("path", {d: "M38,20.5 C38,30.1685093 30.1685093,38 20.5,38"})
            ),
            React.createElement("svg", {className: this.props.classNamespace + "checkmark",
              viewBox: "0 0 70 70"},
              React.createElement("path", {d: "m31.5,46.5l15.3,-23.2"}),
              React.createElement("path", {d: "m31.5,46.5l-8.5,-7.1"})
            ),
            React.createElement("svg", {className: this.props.classNamespace + "cross",
              viewBox: "0 0 70 70"},
              React.createElement("path", {d: "m35,35l-9.3,-9.3"}),
              React.createElement("path", {d: "m35,35l9.3,9.3"}),
              React.createElement("path", {d: "m35,35l-9.3,9.3"}),
              React.createElement("path", {d: "m35,35l9.3,-9.3"})
            )
          )
        )
      );
    },

    getStateClass: function() {
      var className = "workout-progress";
      if (this.props.insightIndex >= this.props.insights.items.length) {
        className += " complete";
      } else {
        className += " not-complete";
      }
      return className;
    },

    loading: function() {
      this.setState({currentState: 'loading'});
    },

    notLoading: function() {
      this.setState({currentState: ''});
    },

    success: function(callback) {
      this.setState({currentState: 'success'});
      this._timeout = setTimeout(function() {
        callback = callback || this.props.onSuccess;
        if (callback && typeof callback === "function") {
          callback();
        }
        this.setState({currentState: ''});
      }.bind(this), this.props.durationSuccess);
    },

    error: function(callback) {
      this.setState({currentState: 'error'});
      this._timeout = setTimeout(function() {
        callback = callback || this.props.onError;
        if (callback && typeof callback === "function") {
          callback();
        }
        this.setState({currentState: ''});
      }.bind(this), this.props.durationError);
    },

    componentWillReceiveProps: function(nextProps) {
      switch(nextProps.state) {
        case 'success':
          this.success();
          return;
        case 'error':
          this.error();
          return;
        case 'loading':
          this.loading();
          return;
        case '':
          this.notLoading();
          return;
        default:
          return;
      }
    },

    componentWillUnmount: function() {
      clearTimeout(this._timeout);
    }
  });

  return ProgressButton;
});
